server/text_generation_server/utils/layers.py

```diff
@@ -509,11 +508,14 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
 
 
 try:
-    import dropout_layer_norm
+    if IS_CUDA_SYSTEM:
+        import dropout_layer_norm
+    else:
+        dropout_layer_norm = None
 
     class FastLayerNorm(nn.LayerNorm):
         def forward(self, hidden_states, residual=None):
-            if hidden_states.shape[-1] > 8192:
+            if hidden_states.shape[-1] > 8192 or IS_ROCM_SYSTEM:
```

Comment:
There is no layer norm implementation in `layernorm_ops`?

Reply:
There is, but I believe it was not updated in the RoCm fork https://github.com/ROCmSoftwarePlatform/flash-attention/tree/flash_attention_for_rocm/csrc/layer_norm and I am just getting errors when trying to compile layernorm/rotary for RoCm:
```
Traceback (most recent call last):
  File "/flash-attention/csrc/layer_norm/setup.py", line 98, in <module>
    raise_if_cuda_home_none("--fast_layer_norm")
  File "/flash-attention/csrc/layer_norm/setup.py", line 48, in raise_if_cuda_home_none
    raise RuntimeError(
RuntimeError: --fast_layer_norm was requested, but nvcc was not found.  Are you sure your environment has nvcc available?  If you're installing within a container from https://hub.docker.com/r/pytorch/pytorch, only images whose names contain 'devel' will provide nvcc.
```

I guess it should be doable to adapt the kernels to make them RoCm compatible though.

Reply:
Didn't you add it for idefics ? Or is it RMS vs Layernorm ?

Reply:
Apology, in my previous message I mixed up `layernorm_ops` & `dropout_layer_norm.dropout_add_ln_fwd`.

Yes indeed, IDEFICS uses RMSNorm (`is_rms_norm=true`), so VLLM's `layernorm_ops.rms_norm` can be used there.

For `layers.py`, we have always `is_rms_norm=false` & VLLM layer norm only implements RMSNorm: https://github.com/vllm-project/vllm/blob/main/csrc/layernorm.cpp. So in `layers.py` we simply dispatch to the torch eager implementation.